scripts/pipelines/log-buildversions.sh

```diff
@@ -0,0 +1,102 @@
+#!/bin/bash
+# This script creates a file documenting the commits
+# used to during the most recent build. The file
+# includes both the commits used by AUTOREV packages
+# and the commits used for the git submodules.
+
+SCRIPT_ROOT=$(realpath $(dirname $BASH_SOURCE))
+
+function usage() {
+    cat <<EOF
+$(basename $0) [-h] [-o FILEPATH]
+
+Get information on the specific commit versions for NILRT components after a build.
+
+# Options
+-h
+    Print this message and exit.
+-o
+    The filepath to output the commit version information to.
+-a
+    Include all SRCREVs and not just AUTOREVs.
+EOF
+}
+
+logfile_path=
+buildhistory_opts=
+while getopts ":ho:a" opt; do
+    case $opt in
+        h )
+            usage >&1
+            exit 0
+            ;;
+        o )
+            if [ ${OPTARG} = "/*" ]; then
+                logfile_path="$OPTARG"
+            else
+                logfile_path="$SCRIPT_ROOT/$OPTARG"
+            fi
+            ;;
+        a )
+            buildhistory_opts="-a"
+            ;;
+        \: )
+            echo "ERROR: Missing argument for option -${OPTARG}" >&2
+            exit 1
+            ;;
+        \? )
+            echo "ERROR: Invalid option -${OPTARG}." >&2
+            exit 1
+            ;;
+    esac
+done
+shift $(($OPTIND - 1))
+
+# Gets nilrt and submodule repo hashes.
+function get_repo_hashes() {
+    local nilrt_rev=$(git rev-parse HEAD)
+    local submodule_status=$(git submodule status)
+    cat <<EOF
+# git repo hashes for nilrt and submodules.
+# Checkout these commits to use the same versions as this build.
+nilrt commit id: $nilrt_rev
+
+# Submodules
+$submodule_status
+EOF
+}
+
+# Gets SRCREV values from the OE buildhistory.
+function get_rev_commits() {
+    local buildhistory_dir=buildhistory/
+    if [ -d buildhistory/ ]; then
+        local buildhistory_output="$(buildhistory-collect-srcrevs ${buildhistory_opts})"
+        cat <<EOF
+# Build revisions used for recipes. Copy into local.conf or a similar
+# global conf file to staple to the same revisions used for this build.
+$buildhistory_output
+EOF
+    else
+        >&2 echo "ERROR: No buildhistory directory at $buildhistory_dir. Make sure a build with buildhistory enabled occured."
+        exit 1
+    fi
+}
+
+. "${SCRIPT_ROOT}/build.common.sh"
+echo
+
+REPO_HASHES="$(get_repo_hashes)"
+ACTUALREVS="$(get_rev_commits)"
+
+if [ -n "$logfile_path" ]; then
+    echo "Writing build version log to $logfile_path."
+    cat > $logfile_path <<EOF
+$REPO_HASHES
+
+$ACTUALREVS
+EOF
```

Comment:
Why have this whole output file argument option? Couldn't the caller redirect stdout to a file, if that's what they want?

Reply:
I honestly hadn't considered that option. Since I'm planning on eventually adding a script to interpret the file this script outputs and apply the hashes, I'd like to keep the script in charge of the file generation. Is that reasonable or is it still unnecessary and just letting the caller redirect better?

Reply:
I'm just suggesting this design because it is more [unixy](https://en.wikipedia.org/wiki/Unix_philosophy#Doug_McIlroy_on_Unix_programming) and I think it would allow you to delete quite a few lines from the script. `stdout` is, conceptually, already the blessed output-file.

>Since I'm planning on eventually adding a script to interpret the file this script outputs and apply the hashes

`log_buildversions.sh | interpret_hashes.sh`

This isn't blocking. So if you don't want to go this route, you can just close this thread.